File: backend/node/models/userModel.js
const { getDB } = require('../db/mongodb');

const createUser = async (username, email, passwordHash) => {
  const db = getDB();
  if (!db) {
    throw new Error('Database not connected');
  }
  
  const user = {
    username,
    email,
    password_hash: passwordHash,
    level: 1,
    xp: 0,
    streak_days: 0,
    college: '',
    battle_rating: 1200,
    battle_wins: 0,
    battle_losses: 0,
    created_at: new Date()
  };
  
  const result = await db.collection('users').insertOne(user);
  return { id: result.insertedId.toString(), ...user, _id: result.insertedId };
};

const findUserByEmail = async (email) => {
  const db = getDB();
  if (!db) {
    return null;
  }
  return await db.collection('users').findOne({ email });
};

const findUserById = async (id) => {
  const db = getDB();
  if (!db) {
    return null;
  }
  const { ObjectId } = require('mongodb');
  try {
    return await db.collection('users').findOne({ _id: new ObjectId(id) });
  } catch (error) {
    console.error('Invalid user ID format:', error.message);
    return null;
  }
};

const updateUserXP = async (userId, xpGained) => {
  const db = getDB();
  if (!db) {
    return { level: 1, xp: 0 };
  }
  const { ObjectId } = require('mongodb');
  const user = await findUserById(userId);
  if (!user) {
    throw new Error('User not found');
  }
  const newXP = (user.xp || 0) + xpGained;
  const newLevel = Math.floor(newXP / 100) + 1;
  
  await db.collection('users').updateOne(
    { _id: new ObjectId(userId) },
    { $set: { xp: newXP, level: newLevel } }
  );
  
  return { level: newLevel, xp: newXP };
};

const updateStreak = async (userId) => {
  const db = getDB();
  if (!db) {
    return { streak_days: 0 };
  }
  const { ObjectId } = require('mongodb');
  const user = await findUserById(userId);
  if (!user) {
    return { streak_days: 0 };
  }
  
  const today = new Date().toDateString();
  const lastActive = user.last_active ? new Date(user.last_active).toDateString() : null;
  
  let streakDays = user.streak_days || 0;
  
  if (lastActive !== today) {
    const yesterday = new Date();
    yesterday.setDate(yesterday.getDate() - 1);
    
    if (lastActive === yesterday.toDateString()) {
      streakDays += 1;
    } else if (lastActive !== today) {
      streakDays = 1;
    }
    
    await db.collection('users').updateOne(
      { _id: new ObjectId(userId) },
      { $set: { streak_days: streakDays, last_active: new Date() } }
    );
  }
  
  return { streak_days: streakDays };
};

const deleteUser = async (userId) => {
  const db = getDB();
  if (!db) {
    throw new Error('Database not connected');
  }
  const { ObjectId } = require('mongodb');
  await db.collection('users').deleteOne({ _id: new ObjectId(userId) });
  return { success: true };
};

module.exports = {
  createUser,
  findUserByEmail,
  findUserById,
  updateUserXP,
  updateStreak,
  deleteUser
};

File: backend/python/models/user_model.py
from db.mongodb import get_db
from bson import ObjectId
from datetime import datetime

def create_user(username, email, password_hash):
    db = get_db()
    user = {
        'username': username,
        'email': email,
        'password_hash': password_hash,
        'level': 1,
        'xp': 0,
        'streak_days': 0,
        'college': '',
        'battle_rating': 1200,
        'battle_wins': 0,
        'battle_losses': 0,
        'created_at': datetime.utcnow()
    }
    result = db.users.insert_one(user)
    user['_id'] = result.inserted_id
    return user

def find_user_by_email(email):
    db = get_db()
    return db.users.find_one({'email': email})

def find_user_by_id(user_id):
    db = get_db()
    return db.users.find_one({'_id': ObjectId(user_id)})

def update_user_xp(user_id, xp_gained):
    db = get_db()
    user = find_user_by_id(user_id)
    new_xp = (user.get('xp', 0) or 0) + xp_gained
    new_level = new_xp // 100 + 1
    db.users.update_one({'_id': ObjectId(user_id)}, {'$set': {'xp': new_xp, 'level': new_level}})
    return {'level': new_level, 'xp': new_xp}

def update_streak(user_id):
    db = get_db()
    user = find_user_by_id(user_id)
    today = datetime.utcnow().date()
    last_active = user.get('last_active')
    streak_days = user.get('streak_days', 0)
    
    if last_active:
        last_date = last_active.date() if isinstance(last_active, datetime) else last_active
        if last_date != today:
            from datetime import timedelta
            yesterday = today - timedelta(days=1)
            if last_date == yesterday:
                streak_days += 1
            else:
                streak_days = 1
    else:
        streak_days = 1
    
    db.users.update_one({'_id': ObjectId(user_id)}, {'$set': {'streak_days': streak_days, 'last_active': datetime.utcnow()}})
    return {'streak_days': streak_days}

def delete_user(user_id):
    db = get_db()
    db.users.delete_one({'_id': ObjectId(user_id)})
    return {'success': True}
